pydantic/_internal/_fields.py

```diff
@@ -246,10 +246,13 @@ def collect_fields(  # noqa: C901
             # 1. To match the behaviour of annotation-only fields
             # 2. To avoid false positives in the NameError check above
             try:
-                delattr(cls, ann_name)
                 if cls.__pydantic_generic_parameters__:  # model can be parametrized
                     assert cls.__pydantic_generic_defaults__ is not None
                     cls.__pydantic_generic_defaults__[ann_name] = default
+                if not (is_dataclass and not isinstance(default, (dataclasses.Field, FieldInfo))):
+                    # for dataclasses we only delete the attribute if it is an instance of
+                    # field, e.g. `a: int = 1` should not get deleted
+                    delattr(cls, ann_name)
```

Comment:
Before this, the following didn't work:

```python
import pydantic

@pydantic.dataclasses.dataclass
class Parent:
    a: int = 1

print(Parent.a)  # AttributeError: type object 'Parent' has no attribute 'a'
```

This fixes it and doesn't break anything else, _but_ the example does currently work for `BaseModel` and I can't tell why. So `BaseModel` must be doing something slightly different, which I don't like because I would like them to both behave the same and share the same implementation in this regard at least.

Reply:
As noted on slack — let's just modify this so that if it is a dataclasses.Field or FieldInfo that has a default, it does `setattr(cls, ann_name, default)` (same as it does in the standard dataclasses logic).

Reply:
[8c98fae](https://github.com/pydantic/pydantic/pull/5276/commits/8c98fae49b0c86de055449b4fe50009750e98ae1)